src/hotspot/share/gc/parallel/parallelScavengeHeap.cpp

```diff
@@ -549,8 +549,26 @@ void ParallelScavengeHeap::collect(GCCause::Cause cause) {
     return;
   }
 
-  VM_ParallelGCSystemGC op(gc_count, full_gc_count, cause);
-  VMThread::execute(&op);
+  while (true) {
+    VM_ParallelGCSystemGC op(gc_count, full_gc_count, cause);
+    VMThread::execute(&op);
+
+    if (!VM_ParallelGCSystemGC::is_cause_full(cause) || op.gc_succeeded()) {
```

Comment:
`!is_cause_full` would cover more cases than `System.gc` and whitebox-fullgc, right? Is this really intended?

The introduce of `op.gc_succeeded()` is not well motivated -- the semantics of the return-val of `invoke()` is also not obvious at first glance. Therefore, I'd prefer keeping the existing signature.

Reply:
>  !is_cause_full would cover more cases than System.gc and whitebox-fullgc, right? Is this really intended?

Yes, it is intended. 

> The introduce of `op.gc_succeeded()` is not well motivated -- the semantics of the return-val of `invoke()` is also not obvious at first glance. Therefore, I'd prefer keeping the existing signature.

Yeah, I did think about this. I can revert it.

Reply:
> Yes, it is intended.

Then, even non-explicit-gc (e.g. `_metadata_GC_threshold` or some other gc-cause) would get "guarantee that at least a Full GC is executed", not matching the title of this ticket.

> Yeah, I did think about this. I can revert it.

Thank you.